File: scala/src/main/scala/clase4/ProducerConsumerStream.scala
package clase4

object ProducerConsumerStream extends App {

  //final class Cons[+A](hd: A, tl: => Stream[A]) extends Stream[A] {
  // `tail` is by name / "lazy"

  def producer(max:Int): Stream[Int] = {

    def doProduce(n:Int): Stream[Int] = {
      if (n < max) {
        println("produzco " + n)
        n #:: doProduce(n + 1)
      }
      else
        Stream.empty
    }

    doProduce(0)

  }

  def consumer(what: Stream[Int]) {
    for(n <- what if (n % 2) == 0) {
      println(n)
    }
  }

  consumer(producer(50))

}

File: haskell/src/clase4/ProducerConsumerStream.hs
module ProducerConsumerStream where

import Control.Concurrent
import Control.Concurrent.Chan

doProduce :: Int -> Int -> [Int]
doProduce n max_ =
  if n < max_ then 
    n : (doProduce (n + 1) max_)
  else 
    []

producer max_ = doProduce 0 max_

consumer (x:xs) = do
  if ((x `mod` 2) == 0) then 
    print x
  else
    return ()
  consumer xs  

consumer [] = return ()

main = consumer(producer(50))
